### PR TITLE
Qsignature add threading to Compare

### PR DESCRIPTION
**Pull request by holmeso**
# Description

The primary use case for the `Compare` class is to compare our `qsignature` vcf files across each study on a regular basis.
Some of our larger studies (hello melanoma) have so many files that to perform the compare in a single-threaded way would take hours.
Fortunately, the components of `qsignatures`'s `Compare` (ie. reading in vcf files and comparing them) scale relatively well when run with multiple threads.

And so, I have incorporated the existing `noOfThreads` option into the `Compare` class.
This creates thread pools of the specified size, enabling additional threads to both read in `qsignature` vcf files and also to perform the comparisons.

Also changed is the collection used to store the genotype values. It was previously `trove`'s `TIntShortMap` and it has now changed to `trove`'s `TIntByteMap`
The motivation for this change was to try and reduce the memory footprint when comparing large numbers of files.
It is possible to store all the different genotype states within a byte primitive and it uses half the bits of a short.

If the user wishes to run this class in a single-threaded way, this is possible by either leaving out the `noOfThreads` option (ensuring backwards compatibility) or by setting the `noOfThreads` option to 1.
 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I have added additional unit tests to check that using a byte gives the same results as using a short to store the genotype information.
I've run in multi-treaded mode against the same data set multiple times and the outputs have been identical. Have also run single-threaded against the same data set with resulted in the same output.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java [hunk 1]
```diff
@@ -201,10 +203,12 @@ private int engage() throws Exception {
 				List<String> rgs = new ArrayList<>(rgResults.getSecond().keySet());
 				for (int i = 0 ; i < rgs.size() ; i++) {
 					String rg1 = rgs.get(i);
-					TIntShortHashMap r1 = rgResults.getSecond().get(rg1);
+					TIntByteHashMap r1 = rgResults.getSecond().get(rg1);
+//					TIntShortHashMap r1 = rgResults.getSecond().get(rg1);
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java]: rm comment

File: qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java [hunk 2]
```diff
@@ -201,10 +203,12 @@ private int engage() throws Exception {
...
 					for (int j = i + 1 ; j < rgs.size() ; j++) {
 						String rg2 = rgs.get(j);
-						TIntShortHashMap r2 = rgResults.getSecond().get(rg2);
+						TIntByteHashMap r2 = rgResults.getSecond().get(rg2);
+//						TIntShortHashMap r2 = rgResults.getSecond().get(rg2);
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java]: ditto

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java [hunk 3]
```diff
@@ -342,17 +345,34 @@ public static TIntShortHashMap loadSignatureRatiosFloatGenotype(File file, int m
...
 				}
 				
-				String[] params = TabTokenizer.tokenize(line);
-				String coverage = params[7];
+				/*
+				 * don't tokenise the line, just get the first 2 columns along with the last
+				 * should be quicker than a full tokenise...
+				 */
+				
+//				String[] params = TabTokenizer.tokenize(line);
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: rm comment

File: qsignature/src/org/qcmg/sig/Compare.java [hunk 4]
```diff
@@ -228,6 +212,106 @@ private void addFilesToMap(List<File> orderedFiles) {
...
+			e.printStackTrace();
+		}
+	}
+	
+	private void populateCache(List<File> files) {
+		AbstractQueue<File> queue =  new ConcurrentLinkedQueue<>(files);
+		ExecutorService service = Executors.newFixedThreadPool(nThreads);		
+		for (int i = 0 ; i < nThreads; i++) {
+			service.execute(() -> {
+					while (true) {
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/Compare.java]: same question

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java [hunk 5]
```diff
@@ -537,11 +525,12 @@ public static void getDataFromBespolkeLayout(File file, int minCoverage, int min
...
 				
-				short g = getCodedGenotype(f);
+				byte genotype1 = getCodedGenotypeAsByte(f);
 				
-				if (isCodedGenotypeValid(g)) {
-					cachePosition.set(ChrPositionCache.getChrPositionIndex(params[0], Integer.parseInt(params[1])));
-					ratios.put(cachePosition.get(), g);
+				if (isCodedGenotypeValid(genotype1)) {
+					cachePosition.set(ChrPositionCache.getStringIndex(params[0] + "\t" + params[1]));
+//					cachePosition.set(ChrPositionCache.getChrPositionIndex(params[0], Integer.parseInt(params[1])));
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: rm

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java [hunk 6]
```diff
@@ -84,6 +87,49 @@ public void testGetPatientFromFile() {
 		assertEquals("UNKNOWN", SignatureUtil.getPatientFromFile(new File("/path/project/AAAA_33/SNP_array/5555.txt")));
 	}
 	
+	@Ignore
```
Comment (delocalizer) [inline qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: new test but ignoring?

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java [hunk 7]
```diff
@@ -289,7 +335,8 @@ public void bespokeGenotype() throws IOException {
 		File f = testFolder.newFile();
 		createBamVcfFile(f);
 		
-		Pair<SigMeta, TMap<String, TIntShortHashMap>> p = SignatureUtil.loadSignatureRatiosBespokeGenotype(f, 10, 10);
+		Pair<SigMeta, TMap<String, TIntByteHashMap>> p = SignatureUtil.loadSignatureGenotype(f, 10, 10);
+//		Pair<SigMeta, TMap<String, TIntShortHashMap>> p = SignatureUtil.loadSignatureRatiosBespokeGenotype(f, 10, 10);
```
Comment (delocalizer) [inline qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: rm

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java [hunk 8]
```diff
@@ -344,10 +391,14 @@ public void compareBams() throws IOException {
 		File f2 = testFolder.newFile();
 		createBamVcfFile(f1);
 		createBamVcfFile(f2);
-		Pair<SigMeta, TMap<String, TIntShortHashMap>> p1 = SignatureUtil.loadSignatureRatiosBespokeGenotype(f1,10,10);
-		Pair<SigMeta, TMap<String, TIntShortHashMap>> p2 = SignatureUtil.loadSignatureRatiosBespokeGenotype(f2,10,10);
-		Pair<SigMeta, TIntShortHashMap> ratios1 = new Pair<>(p1.getKey(), p1.getSecond().get("all"));
-		Pair<SigMeta, TIntShortHashMap> ratios2 = new Pair<>(p2.getKey(), p2.getSecond().get("all"));
+		Pair<SigMeta, TMap<String, TIntByteHashMap>> p1 = SignatureUtil.loadSignatureGenotype(f1,10,10);
+		Pair<SigMeta, TMap<String, TIntByteHashMap>> p2 = SignatureUtil.loadSignatureGenotype(f2,10,10);
+//		Pair<SigMeta, TMap<String, TIntShortHashMap>> p1 = SignatureUtil.loadSignatureGenotype(f1,10,10);
```
Comment (delocalizer) [inline qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: ditto (and all commented tests below?)

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java [hunk 9]
```diff
@@ -580,65 +569,71 @@ public static void getDataFromBespolkeLayout(File file, int minCoverage, int min
...
-	 * If over 90% in C, then will return 200
-	 * If over 90% in G, then will return 20
-	 * If over 90% in T, then will return 2
-	 * If het for AC, will return 1100
-	 * If het for AG, will return 1010
-	 * If het for AT, will return 1001
-	 * If het for CG, will return 0110
-	 * If het for CT, will return 0101
-	 * If het for GT, will return 0011
+	 * If over 90% in A, then will return 1000000 (binary form)
```
Comment (delocalizer) [inline qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: I think it'd be good to have this mapping between ints and well-defined genotypes explicit somewhere as static finals so there aren't so many magic numbers floating about here and in tests

File: qcommon/src/org/qcmg/common/util/NumberUtils.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	 * 
+	 * Try and minimise the number of calls to Arrays.binarySearch by checking size and boundary positions
+	 * Should not be more than 500 entries in the array 
+	 * 
+	 * 
+	 * @param array
+	 * @param value
+	 * @return
+	 */
+	public static int getPositionOfLongInArray(long [] array, long value) {
```
[the remark this answers is not in the document]
Comment (holmeso) [reply qcommon/src/org/qcmg/common/util/NumberUtils.java]: Yeah, in my testing it didn't really do much - will remove

File: qsignature/src/org/qcmg/sig/Compare.java [hunk 11]
```diff
@@ -228,6 +212,106 @@ private void addFilesToMap(List<File> orderedFiles) {
...
+		AbstractQueue<Integer> queue =  new ConcurrentLinkedQueue<>();
+		for (int i = 0 ; i < size -1 ; i++) {
+			queue.add(i);
+		}
+		
+		ExecutorService service = Executors.newFixedThreadPool(nThreads);		
+		for (int i = 0 ; i < nThreads; i++) {
+			service.execute(() -> {
+					List<Comparison> myComps = new ArrayList<>();
+					while (true) {
```
[the remark this answers is not in the document]
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/Compare.java]: switched

File: qcommon/test/org/qcmg/common/util/NumberUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+import org.junit.Test;
+
+public class NumberUtilsTest {
+	
+	@Test
+	public void pack2IntsInto1() {
+		assertEquals(0, NumberUtils.pack2IntsInto1(0, 0));
+		assertEquals(1, NumberUtils.pack2IntsInto1(0, 1));
+		assertEquals(1 << 16, NumberUtils.pack2IntsInto1(1, 0));
```
[the remark this answers is not in the document]
Comment (holmeso) [reply qcommon/test/org/qcmg/common/util/NumberUtilsTest.java]: yep

File: qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java (hunk 1 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java]: done

File: qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java (hunk 2 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/CompareGenotypeBespoke.java]: ditto

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java (hunk 6 above)
Comment (holmeso) [reply qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: have deleted - just used to test speed

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java (hunk 7 above)
Comment (holmeso) [reply qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: done

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java (hunk 9 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: I agree - done

File: qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java (hunk 8 above)
Comment (holmeso) [reply qsignature/test/org/qcmg/sig/util/SignatureUtilTest.java]: done

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java (hunk 5 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: yep

File: qsignature/src/org/qcmg/sig/Compare.java (hunk 4 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/Compare.java]: done

File: qsignature/src/org/qcmg/sig/util/SignatureUtil.java (hunk 3 above)
Comment (holmeso) [reply qsignature/src/org/qcmg/sig/util/SignatureUtil.java]: gone

**Comment by holmeso** [comment]
Thanks for your review

**Comment by delocalizer** [comment]
> Thanks for your review

No worries; a privilege and an education as always.

**Pull request merged**